vms/avm/network/gossip.go

```diff
@@ -0,0 +1,162 @@
+	"github.com/ava-labs/avalanchego/vms/avm/txs/mempool"
+)
+
+var (
+	_ p2p.Handler                = (*txGossipHandler)(nil)
+	_ gossip.Set[*txs.Tx]        = (*gossipMempool)(nil)
+	_ gossip.Marshaller[*txs.Tx] = (*txParser)(nil)
+)
+
+// txGossipHandler is the handler called when serving gossip messages
+type txGossipHandler struct {
+	p2p.NoOpHandler
+	appGossipHandler  p2p.Handler
+	appRequestHandler p2p.Handler
+}
+
+func (t txGossipHandler) AppGossip(
+	ctx context.Context,
+	nodeID ids.NodeID,
+	gossipBytes []byte,
+) {
+	t.appGossipHandler.AppGossip(ctx, nodeID, gossipBytes)
+}
+
+func (t txGossipHandler) AppRequest(
+	ctx context.Context,
+	nodeID ids.NodeID,
+	deadline time.Time,
+	requestBytes []byte,
+) ([]byte, error) {
+	return t.appRequestHandler.AppRequest(ctx, nodeID, deadline, requestBytes)
+}
+
+type txParser struct {
+	parser txs.Parser
+}
+
+func (*txParser) MarshalGossip(tx *txs.Tx) ([]byte, error) {
+	return tx.Bytes(), nil
+}
+
+func (g *txParser) UnmarshalGossip(bytes []byte) (*txs.Tx, error) {
+	return g.parser.ParseTx(bytes)
+}
+
+func newGossipMempool(
+	mempool mempool.Mempool,
+	log logging.Logger,
+	txVerifier TxVerifier,
+	parser txs.Parser,
+	maxExpectedElements uint64,
+	falsePositiveProbability,
+	maxFalsePositiveProbability float64,
+) (*gossipMempool, error) {
+	bloom, err := gossip.NewBloomFilter(maxExpectedElements, falsePositiveProbability)
+	return &gossipMempool{
+		Mempool:                     mempool,
+		log:                         log,
+		txVerifier:                  txVerifier,
+		parser:                      parser,
+		maxFalsePositiveProbability: maxFalsePositiveProbability,
+		bloom:                       bloom,
+	}, err
+}
+
+type gossipMempool struct {
+	mempool.Mempool
+	log                         logging.Logger
+	txVerifier                  TxVerifier
+	parser                      txs.Parser
+	maxFalsePositiveProbability float64
+
+	lock  sync.RWMutex
+	bloom *gossip.BloomFilter
+}
+
+// Add is called by the p2p SDK when handling transactions that were pushed to
+// us and when handling transactions that were pulled from a peer. If this
+// returns a nil error while handling push gossip, the p2p SDK will queue the
+// transaction to push gossip as well.
+func (g *gossipMempool) Add(tx *txs.Tx) error {
+	txID := tx.ID()
+	if _, ok := g.Mempool.Get(txID); ok {
+		return fmt.Errorf("attempted to issue %w: %s ", mempool.ErrDuplicateTx, txID)
+	}
+
+	if reason := g.Mempool.GetDropReason(txID); reason != nil {
+		// If the tx is being dropped - just ignore it
+		//
+		// TODO: Should we allow re-verification of the transaction even if it
+		// failed previously?
+		return reason
+	}
+
+	// Verify the tx at the currently preferred state
+	if err := g.txVerifier.VerifyTx(tx); err != nil {
+		g.Mempool.MarkDropped(txID, err)
+		return err
+	}
+
+	return g.AddVerified(tx)
+}
+
+func (g *gossipMempool) AddVerified(tx *txs.Tx) error {
+	if err := g.Mempool.Add(tx); err != nil {
```

Comment:
Should we check if `tx` is already in the mempool?

Reply:
`AddVerified` is expected to error if the tx is already in the mempool (which is what happens here)